clients/azuredevopsrepo/zip.go

```diff
@@ -0,0 +1,251 @@
+	}
+
+	if _, err := io.Copy(repoFile, res.Body); err != nil {
+		return fmt.Errorf("io.Copy: %w", err)
+	}
+
+	z.tempDir = tempDir
+	z.tempZipFile = repoFile.Name()
+
+	return nil
+}
+
+func (z *zipHandler) getLocalPath() (string, error) {
+	if err := z.setup(); err != nil {
+		return "", fmt.Errorf("error during zipHandler.setup: %w", err)
+	}
+	absTempDir, err := filepath.Abs(z.tempDir)
+	if err != nil {
+		return "", fmt.Errorf("error during filepath.Abs: %w", err)
+	}
+	return absTempDir, nil
+}
+
+func (z *zipHandler) extractZip() error {
+	in, err := os.OpenFile(z.tempZipFile, os.O_RDONLY, 0o644)
+	if err != nil {
+		return fmt.Errorf("os.OpenFile: %w", err)
+	}
+
+	stat, err := in.Stat()
+	if err != nil {
+		return fmt.Errorf("os.File.Stat: %w", err)
+	}
+
+	zipReader, err := zip.NewReader(in, stat.Size())
+	if err != nil {
+		return fmt.Errorf("zip.NewReader: %w", err)
+	}
+
+	z.files = make([]string, 0, len(zipReader.File))
+	for _, file := range zipReader.File {
+		//nolint:gosec // G305: Handling of file paths is done below
+		filenamepath := filepath.Join(z.tempDir, file.Name)
+		if !strings.HasPrefix(filepath.Clean(filenamepath), filepath.Clean(z.tempDir)) {
+			return errInvalidFilePath
+		}
+		if file.FileInfo().IsDir() {
+			if err := os.MkdirAll(filenamepath, 0o755); err != nil {
+				return fmt.Errorf("error during os.MkdirAll: %w", err)
+			}
+			continue
+		}
+
+		outFile, err := os.OpenFile(filenamepath, os.O_CREATE|os.O_WRONLY, 0o644)
+		if err != nil {
+			return fmt.Errorf("os.OpenFile: %w", err)
+		}
+
+		rc, err := file.Open()
+		if err != nil {
+			return fmt.Errorf("file.Open: %w", err)
+		}
+
+		written, err := io.CopyN(outFile, rc, maxSize)
+		if err != nil && !errors.Is(err, io.EOF) {
+			return fmt.Errorf("%w io.Copy: %w", errZipNotFound, err)
+		}
+		if written >= maxSize {
+			return errFileTooLarge
+		}
+		outFile.Close()
+
+		filename := strings.TrimPrefix(filenamepath, filepath.Clean(z.tempDir)+string(os.PathSeparator))
+		z.files = append(z.files, filename)
+	}
+	return nil
+}
+
+func (z *zipHandler) listFiles(predicate func(string) (bool, error)) ([]string, error) {
+	if err := z.setup(); err != nil {
+		return nil, fmt.Errorf("error during zipHandler.setup: %w", err)
+	}
+	ret := make([]string, 0)
+	for _, file := range z.files {
+		matches, err := predicate(file)
+		if err != nil {
+			return nil, err
+		}
+		if matches {
+			ret = append(ret, file)
+		}
+	}
+	return ret, nil
+}
+
+func (z *zipHandler) getFile(filename string) (*os.File, error) {
+	if err := z.setup(); err != nil {
+		return nil, fmt.Errorf("error during zipHandler.setup: %w", err)
+	}
+	f, err := os.Open(filepath.Join(z.tempDir, filename))
+	if err != nil {
+		return nil, fmt.Errorf("open file: %w", err)
+	}
+	return f, nil
+}
+
+func (z *zipHandler) cleanup() error {
+	if err := os.RemoveAll(z.tempDir); err != nil && !os.IsNotExist(err) {
+		return fmt.Errorf("os.Remove: %w", err)
+	}
+
+	z.files = nil
+	return nil
+}
```

Comment:
Does it make sense to call this from within the toplevel Client.Close()?
https://github.com/ossf/scorecard/blob/f4d721d307d53e54346aa165b796e5ecfa5460cb/clients/azuredevopsrepo/client.go#L185-L187

Reply:
Did you mean to push a commit before resolving this ?

Reply:
Ah, sorry. I made the change locally, but didn't push.